### PR TITLE
Add Seeker interface to file

### PR DESCRIPTION
**Pull request by HolzeHan**
This PR implements the optional seeker interface for file

### CODE REVIEW

**Comment by jszwec** [comment]
Before I start reviewing, my first question is - why? What's the use case?

Each Seek call essentially makes another call to s3 which costs you money. There are a lot of functions that can take advantage of seeker behind the scenes, potentially making your aws bill higher.

It also introduces some issues, such as you not closing the body while calling seek (if it's opened) thus leaking go routines.

**Comment by HolzeHan** [comment]
Currently I am streaming 10GB files from s3 into an application. The file has a streamable format, so it can be read from most positions. It would therefore speed up my processing time, if I could seek in the file.
In addition I want to add a wrapper around a fs which allows to stream data using multipart download. For that I need the fs to support the seek operation (every part needs to seek to its starting position and read from there). This would speed up the streaming from around 30MB/s to somewhere around 200MB/s

Leaking go routines is a good point. I'll check again with that in mind

File: file.go [hunk 1]
```diff
@@ -10,14 +14,104 @@ import (
...
+			Key:    aws.String(f.name),
+			Range:  aws.String(fmt.Sprintf("bytes=%d-", newPosition)),
+		})
+
+	if err != nil {
+		return f.currentPosition, err
+	}
+
+	f.currentPosition = newPosition
+	f.realReader = rawObject.Body
```
Comment (guenhter) [inline file.go:109]: Close the existing reader (if there is any)

File: file.go [hunk 2]
```diff
@@ -10,14 +14,104 @@ import (
...
+		newPosition += offset
+	case io.SeekEnd:
+		stat, err := f.Stat()
+		if err != nil {
+			return 0, err
+		}
+		newPosition = stat.Size() + offset
+	default:
+		return 0, fmt.Errorf("unknown 'whence': %d", whence)
+	}
```
Comment (guenhter) [inline file.go:91]: Assert that the position is correct (e.g. is not -1 or so in case a wrong offset/whence is defined)

File: file.go [hunk 3]
```diff
@@ -10,14 +14,104 @@ import (
...
+	cl     s3iface.S3API
+	bucket string
+	name   string
+
+	realReader      io.ReadCloser
+	stat            func() (fs.FileInfo, error)
+	currentPosition int64
+}
+
+func NewFile(cl s3iface.S3API, bucket string, name string) (fs.File, error) {
```
Comment (guenhter) [inline file.go:30]: I guess this function doesn't create a file but opens one, doesn't it?

File: file.go [hunk 4]
```diff
@@ -1,6 +1,10 @@
 package s3fs
 
 import (
+	"fmt"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/s3"
+	"github.com/aws/aws-sdk-go/service/s3/s3iface"
```
Comment (jszwec) [inline file.go:7]: please fix import order

File: file.go [hunk 5]
```diff
@@ -10,14 +14,104 @@ import (
...
 )
 
 type file struct {
-	io.ReadCloser
-	stat func() (fs.FileInfo, error)
+	cl     s3iface.S3API
+	bucket string
+	name   string
+
+	realReader      io.ReadCloser
```
Comment (jszwec) [inline file.go:25]: please revert and leave `io.ReadCloser` embedded - you don't need to implement `Close` then

File: file.go [hunk 6]
```diff
@@ -10,14 +14,104 @@ import (
...
+	stat            func() (fs.FileInfo, error)
+	currentPosition int64
```
Comment (jszwec) [inline file.go:27]: please rename to `offset` or `i` for index

File: file.go (hunk 3 above)
Comment (jszwec) [reply file.go:30]: This doesn't have to be exported. 

Either revert it and leave it as it was in `Open` 
or
change signature to 
```go
func newFile(cl s3iface.S3API, bucket string, name string) (*file, error)
```

File: file.go [hunk 7]
```diff
@@ -10,14 +14,104 @@ import (
...
+}
+
+func (f *file) Read(p []byte) (int, error) {
+	n, err := f.realReader.Read(p)
+	f.currentPosition += int64(n)
+	return n, err
+}
+
+func (f *file) Close() error {
+	return f.realReader.Close()
```
Comment (jszwec) [inline file.go:73]: if embedded then this is redundant

File: file.go [hunk 8]
```diff
@@ -10,14 +14,104 @@ import (
...
+	case io.SeekCurrent:
+		newPosition += offset
+	case io.SeekEnd:
+		stat, err := f.Stat()
+		if err != nil {
+			return 0, err
+		}
+		newPosition = stat.Size() + offset
+	default:
+		return 0, fmt.Errorf("unknown 'whence': %d", whence)
```
Comment (jszwec) [inline file.go:90]: let's use similar message to bytes.Bytes' -> https://cs.opensource.google/go/go/+/refs/tags/go1.19:src/bytes/reader.go;l=127

File: file.go (hunk 1 above)
Comment (jszwec) [reply file.go:109]: Body will never be nil at this point, so yeah, we can just Close it

File: file.go [hunk 9]
```diff
@@ -10,14 +14,104 @@ import (
...
+	case io.SeekStart:
+		newPosition = offset
+	case io.SeekCurrent:
+		newPosition += offset
+	case io.SeekEnd:
+		stat, err := f.Stat()
+		if err != nil {
+			return 0, err
+		}
+		newPosition = stat.Size() + offset
```
Comment (jszwec) [inline file.go:88]: is there a point to continue and call GetObject in this case?

we probably should implement eofReader and wrap our body into it and call it a day

```go
type eofReader struct {}

func (eofReader) Read([]byte) (int64, error) { return 0, io.EOF }
```

```go
f.ReadCloser = ioutil.NopCloser(eofReader{})
```

File: file.go [hunk 10]
```diff
@@ -10,14 +14,104 @@ import (
...
+	}
+
+	f.currentPosition = newPosition
+	f.realReader = rawObject.Body
+
+	return f.currentPosition, nil
 }
 
-func (f file) Stat() (fs.FileInfo, error) { return f.stat() }
+func (f *file) Stat() (fs.FileInfo, error) { return f.stat() }
```
Comment (jszwec) [inline file.go:114]: why change this?

File: fs.go [hunk 11]
```diff
@@ -4,13 +4,11 @@ package s3fs
...
 import (
 	"errors"
-	"io/fs"
-	"path"
-
 	"github.com/aws/aws-sdk-go/aws"
 	"github.com/aws/aws-sdk-go/aws/awserr"
 	"github.com/aws/aws-sdk-go/service/s3"
 	"github.com/aws/aws-sdk-go/service/s3/s3iface"
+	"io/fs"
```
Comment (jszwec) [inline fs.go:11]: please fix import order

**Review by jszwec — CHANGES_REQUESTED** [review]
I will take a look at the tests once other fixes are made

File: file.go (hunk 1 above)
Comment (HolzeHan) [reply file.go]: fixed. Error of closer will be pushed upwards

File: file.go (hunk 2 above)
Comment (HolzeHan) [reply file.go:94]: done

File: file.go (hunk 3 above)
Comment (HolzeHan) [reply file.go]: right. fixed

File: file.go (hunk 4 above)
Comment (HolzeHan) [reply file.go]: done

File: file.go (hunk 5 above)
Comment (HolzeHan) [reply file.go]: true. done

File: file.go (hunk 6 above)
Comment (HolzeHan) [reply file.go]: renamed

File: file.go (hunk 7 above)
Comment (HolzeHan) [reply file.go]: removed

File: file.go (hunk 8 above)
Comment (HolzeHan) [reply file.go]: done. Hope that is what you meant

File: file.go (hunk 9 above)
Comment (HolzeHan) [reply file.go]: good point. Thanks for the snippet

File: file.go (hunk 10 above)
Comment (HolzeHan) [reply file.go]: You are right. That's not needed. reverted

File: fs.go (hunk 11 above)
Comment (HolzeHan) [reply fs.go]: done

File: file.go [hunk 12]
```diff
@@ -1,20 +1,135 @@
...
+		bucket:     bucket,
+		name:       name,
+		ReadCloser: out.Body,
+		stat:       statFunc,
+		offset:     0,
+	}, nil
+}
+
+func (f *file) Read(p []byte) (int, error) {
+	n, err := f.Read(p)
```
Comment (jszwec) [inline file.go:71]: bug: this will cause stackoverflow

```go
f.ReadCloser.Read(p)
```

File: fs_test.go [hunk 13]
```diff
@@ -44,6 +45,186 @@ func TestMain(m *testing.M) {
...
+
+	t.Run("seek once", func(t *testing.T) {
+		fixtures := []struct {
+			desc     string
+			offset   int64
+			whence   int
+			expected int64
+			err      error
+		}{
+			{"whence SeekStart ", 2, io.SeekStart, 2, nil},
```
Comment (jszwec) [inline fs_test.go:75]: there are too many arguments and it's hard to follow, can we please do the following?

```go
			{
				desc:     "whence SeekStart ",
				offset:   2,
				whence:   io.SeekStart,
				expected: 2,
				err:      nil,
			},
```

Please apply the same for all other tests

File: fs_test.go [hunk 14]
```diff
@@ -44,6 +45,186 @@ func TestMain(m *testing.M) {
...
+			{"whence SeekCurrent", 4, io.SeekCurrent, 4, nil},
+			{"whence SeekEnd", -1, io.SeekEnd, int64(len(content)) - 1, nil},
+			{"whence SeekEnd", 4, 3, 0, fmt.Errorf("unknown 'whence': 3")},
```
Comment (jszwec) [inline fs_test.go:78]: this needs to be updated, please use `errors.New()`

I am not a fan of comparing errors by their message... but I guess it's okay for something as simple

File: fs_test.go [hunk 15]
```diff
@@ -44,6 +45,186 @@ func TestMain(m *testing.M) {
...
+
+				_, err = data.(io.Seeker).Seek(f.offset, f.whence)
+				if err == nil {
+					t.Fatal(err)
+				}
+
+				errorType, ok := err.(awserr.RequestFailure)
+
+				if !ok {
+					t.Errorf("Expected awserr.RequestFailure, got %v", errorType)
```
Comment (jszwec) [inline fs_test.go:131]: this seems wrong

shouldn't it be:
```go
t.Errorf("expected awserr.RequestFailure, got %T", err)
```

and `errorType` is misleading because it is always of type `awserr.RequestFailure`

File: fs_test.go [hunk 16]
```diff
@@ -44,6 +45,186 @@ func TestMain(m *testing.M) {
...
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				_, err = data.(io.Seeker).Seek(f.offset, f.whence)
+				if err == nil {
+					t.Fatal(err)
+				}
+
+				errorType, ok := err.(awserr.RequestFailure)
```
Comment (jszwec) [inline fs_test.go:128]: please use `errors.As()` instead, because we may want to wrap errors to contain additional information

File: fs_test.go [hunk 17]
```diff
@@ -44,6 +45,186 @@ func TestMain(m *testing.M) {
...
+			desc      string
+			readBytes int
+			offset    int64
+			whence    int
+			expected  []byte
+		}{
+			{"whence SeekStart", 3, 2, io.SeekStart, content[2:5]},
+			{"whence SeekStart overflow", 3, 5, io.SeekStart, content[5:7]},
+			{"whence SeekCurrent", 3, 4, io.SeekCurrent, content[4:7]},
+			{"whence SeekEnd", 3, -3, io.SeekEnd, content[len(content)-3:]},
```
Comment (jszwec) [inline fs_test.go:193]: please add following test cases

```go
{"please fill out", 100, 0, io.SeekStart, []byte("content")},
{"please fill out", 100, 1, io.SeekStart, []byte("ontent")},
{"please fill out", 100, 1, io.SeekCurrent, []byte("ontent")},
{"please fill out", 1, 1, io.SeekCurrent, []byte("o")},
```

File: fs_test.go [hunk 18]
```diff
@@ -44,6 +45,186 @@ func TestMain(m *testing.M) {
...
+				readBytes, err := data.Read(readBuffer)
+				if readBytes != len(f.expected) {
+					t.Errorf("Read returned unexpected number of bytes")
+				}
+				if err != nil && err != io.EOF {
+					t.Fatal(err)
+				}
+				if bytes.Compare(readBuffer[:readBytes], f.expected) != 0 {
+					t.Errorf("Expected %s, got %s", string(f.expected), string(readBuffer))
+				}
```
Comment (jszwec) [inline fs_test.go:220]: I am paranoid so let's parallel this with `bytes.Reader` to be sure that our seeker returns the same

```go
				readSeekers := []struct {
					desc   string
					seeker io.ReadSeeker
				}{
					{desc: "bytes reader", seeker: bytes.NewReader(content)},
					{desc: "file", seeker: data.(io.ReadSeeker)},
				}

				for _, rs := range readSeekers {
					t.Run(f.desc, func(t *testing.T) {
						_, err = rs.seeker.Seek(f.offset, f.whence)
						if err != nil {
							t.Fatal(err)
						}

						readBuffer := make([]byte, f.readBytes)
						readBytes, err := rs.seeker.Read(readBuffer)
						if readBytes != len(f.expected) {
							t.Errorf("Read returned unexpected number of bytes")
						}
						if err != nil && err != io.EOF {
							t.Fatal(err)
						}
						if bytes.Compare(readBuffer[:readBytes], f.expected) != 0 {
							t.Errorf("Expected %s, got %s", string(f.expected), string(readBuffer))
						}
					})
				}

```

File: fs_test.go [hunk 19]
```diff
@@ -44,6 +45,186 @@ func TestMain(m *testing.M) {
...
+
+			})
+		}
+	})
+
+}
+
```
Comment (jszwec) [inline fs_test.go:227]: I think we are missing a tes where we seek to the end and read - we should be getting EOF

File: file.go [hunk 20]
```diff
@@ -1,20 +1,135 @@
...
+	// If the position has not moved, there is no need to make a new query
+	if f.offset == newOffset {
+		return newOffset, nil
+	}
+
+	if newOffset < 0 {
+		return f.offset, errors.New("s3fs.file.Seek: seeked to a negative position")
+	}
+	if newOffset > size {
+		return f.offset, errors.New("s3fs.file.Seek: seeked to a position exceeding size")
```
Comment (jszwec) [inline file.go:105]: `bytes.Reader` accepts this case, it only returns the desired offset - https://cs.opensource.google/go/go/+/refs/tags/go1.19:src/bytes/reader.go;l=116
it would return EOF on the next read. I believe we should do the same. We can remove this condition and change line 107
```go
if newOffset == size {
```
to 
```go
if newOffset >= size {
```

we are missing a test case for this, it would be worth adding one

File: file.go [hunk 21]
```diff
@@ -1,20 +1,135 @@
...
+	}
+	if newOffset == size {
+		f.ReadCloser = ioutil.NopCloser(eofReader{})
+		f.offset = newOffset
+		return f.offset, nil
+	}
+
+	err = f.Close()
```
Comment (jszwec) [inline file.go:113]: 1. should f.Close() be called **after** we successfully got the new object?

2. assuming that s3 client uses regular http.Response, there is something to think about - according to http.Response's documentation:
```
        // The default HTTP client's Transport may not
	// reuse HTTP/1.x "keep-alive" TCP connections if the Body is
	// not read to completion and closed.
```
we definitely can't read the entire body if the file is large, but maybe we should determine if the remaining body is small enough and read it to ioutil.Discard to release the connection

**Review by jszwec — CHANGES_REQUESTED** [review]

File: file.go (hunk 12 above)
Comment (guenhter) [reply file.go:71]: Maybe you can also double check the tests. They should find such things.

Comment (jszwec) [reply file.go]: Tests are fine, they panic

File: file.go [hunk 22]
```diff
@@ -1,20 +1,132 @@
...
+	// If the position has not moved, there is no need to make a new query
+	if f.offset == newOffset {
+		return newOffset, nil
+	}
+
+	if newOffset < 0 {
+		return 0, errors.New("s3fs.file.Seek: seeked to a negative position")
+	}
+	if newOffset >= size {
+		f.ReadCloser = ioutil.NopCloser(eofReader{})
```
Comment (jszwec) [inline file.go:105]: Bug: current reader not closed here

**Review by jszwec — CHANGES_REQUESTED** [review]

File: file.go [hunk 23]
```diff
@@ -1,20 +1,132 @@
...
+		f.offset = newOffset
+		return f.offset, nil
+	}
+
+	err = f.Close()
+	if err != nil {
+		return f.offset, err
+	}
+
+	rawObject, err := f.cl.GetObject(
```
Comment (jszwec) [inline file.go:115]: There is one more potential big issue here.

With your current code, cached file info (stat func) may no longer be valid. For example, if in the meantime your file has been overwritten and its size changed, your file's size is no longer valid. 

This is fine with os.File, because it points to the same inode, but that's not the case here.

How would you solve this?

**Review by jszwec — CHANGES_REQUESTED** [review]

File: fs_test.go (hunk 18 above)
Comment (WilliamFrei) [reply fs_test.go]: I tried this, it does not work immediately as the `Read()` behaviour is different between the two:
The bytes.Reader returns io.EOF on a Read() call only when there is nothing left to read and there are no bytes being read.
The ReadCloser that comes back from s3 returns io.EOF on a Read() call if the offset is past the last readable byte after filling the buffer.

To illustrate this with an example:
8 bytes left to read, buffer of size 16:
bytes.Reader reads 8 bytes and returns `8, nil` on the first Read call and `0, io.EOF` on further ones.
"Our" ReadCloser reads 8 bytes and returns `8, io.EOF` on the first call already.

According to this, both are acceptable behaviours:
https://pkg.go.dev/io#Reader
> An instance of this general case is that a Reader returning a non-zero number of bytes at the end of the input stream may return either err == EOF or err == nil. The next Read should return 0, EOF.

As far as I can see, there are 4 ways to solve this:
1. Adapt the `Read()` method in file.go to mimic bytes.Reader behaviour.
2. Don't do the parallel tests for cases where one of the two readers would return io.EOF and the other wouldn't.
3. Add an additional `Read()` call to the tests that read till end of file, to compare the error of this second call, because that call should return `0, io.EOF` for both readers. Ignore differences in errors on the first call, if the error is one of the two valid ones (nil or io.EOF).
4. Distinguish between the two readers in the tests if we expect an EOF, which would be very ugly.

I have chosen 3. for now

(This is by the way not a thing that has changed with the implementation of the Seeker interface - the behaviour of the two readers was already different before)

File: fs_test.go (hunk 19 above)
Comment (WilliamFrei) [reply fs_test.go:524]: Added

File: fs_test.go (hunk 17 above)
Comment (WilliamFrei) [reply fs_test.go]: Added

File: fs_test.go (hunk 14 above)
Comment (WilliamFrei) [reply fs_test.go]: I'm still comparing errors by their message, but those tests are in their own test-subgroup now (starting from line 115).

File: file.go (hunk 20 above)
Comment (WilliamFrei) [reply file.go]: Changed, added tests for both == and >=.

File: file.go (hunk 21 above)
Comment (WilliamFrei) [reply file.go:111]: 1. Changed, closing now happens before any operations that would assign a new value to f.ReadCloser.
2. I'm not quite sure how to implement this. What is small enough? 100 bytes? 1kB?

File: file.go (hunk 22 above)
Comment (WilliamFrei) [reply file.go]: Fixed, closing the reader before this now.

File: file.go (hunk 23 above)
Comment (WilliamFrei) [reply file.go:122]: The current solution is to reassign f.stat (line 135). However, the newly created `getStatFunc` is not very nice, and I would be happy about suggestions how to change it; one reason why it isn't very nice is the fact that I have to pass the current offset to properly calculate the file size (as s3ObjOutput.ContentLength is only the remaining file size).

Comment (jszwec) [reply file.go:122]: Right, but stat in itself is not the real problem. It's the fact that your file could have been overwritten and is now different

File: fs_test.go (hunk 18 above)
Comment (jszwec) [reply fs_test.go]: our test should not be affected by the behavior of the reader (for as long as it's correct)

wouldn't using `io.CopyN` solve this? I implemented this and all tests are passing

```go
				readSeekers := []struct {
					desc   string
					seeker io.ReadSeeker
				}{
					{desc: "bytes reader", seeker: bytes.NewReader(content)},
					{desc: "file", seeker: data.(io.ReadSeeker)},
				}

				for _, rs := range readSeekers {
					t.Run(f.desc, func(t *testing.T) {
						_, err = rs.seeker.Seek(f.offset, f.whence)
						if err != nil {
							t.Fatal(err)
						}

						var buf bytes.Buffer
						_, err := io.CopyN(&buf, rs.seeker, int64(f.readBytes))

						if err != nil {
							t.Fatal(err)
						}

						if buf.String() != string(f.expected) {
							t.Errorf("Expected %s, got %s", string(f.expected), buf.String())
						}
					})
				}
```

File: file.go (hunk 23 above)
Comment (jszwec) [reply file.go:122]: I wrote this little app to illustrate the problem: https://gist.github.com/jszwec/f314aa464d1643c55bb84c689368aee8

If you want to easily start it, use one of the following docker compose files before running it:
https://github.com/jszwec/s3fs/blob/main/test/localstack/docker-compose.yml
https://github.com/jszwec/s3fs/blob/main/test/minio/docker-compose.yml

Imagine what would happen if you used this implementation of seeker with your downloader and the file would change in the meantime...

This is a potential blocker for this PR, because correctness is more important than speed. At the very least we may make the seeker implementation optional, otherwise we may break existing code.

File: fs_test.go (hunk 18 above)
Comment (WilliamFrei) [reply fs_test.go]: Using this now.

File: file.go (hunk 23 above)
Comment (jszwec) [reply file.go:130]: `NewSeekable` sounds like Java, I don't like it. Instead let's use the `Option` design

```go
type Option func(*S3FS)

func WithReadSeeker(fsys *S3FS) { fsys.readSeeker = true }

func New(cl s3iface.S3API, bucket string, opts ...Option) *S3FS {
	fsys := &S3FS{
		cl:     cl,
		bucket: bucket,
	}
	
	for _, opt := range opts {
	  opt(fsys)
	}
	return fsys
}

type S3FS struct {
	cl     s3iface.S3API
	bucket string
	readSeeker bool
}
```

File: file.go [hunk 24]
```diff
@@ -1,20 +1,153 @@
...
+		err, ok := err.(awserr.RequestFailure)
+		if ok && err.StatusCode() == http.StatusPreconditionFailed {
+			return 0, fs.ErrNotExist
+		}
+		return 0, err
+	}
+
+	f.offset = newOffset
+	f.ReadCloser = rawObject.Body
+	f.stat = getStatFunc(f.cl, f.bucket, f.name, *rawObject, f.offset)
```
Comment (jszwec) [inline file.go:148]: is this necessary, since we depend on the same etag?

File: file.go [hunk 25]
```diff
@@ -1,20 +1,153 @@
...
+	rawObject, err := f.cl.GetObject(
+		&s3.GetObjectInput{
+			Bucket:  aws.String(f.bucket),
+			Key:     aws.String(f.name),
+			Range:   aws.String(fmt.Sprintf("bytes=%d-", newOffset)),
+			IfMatch: aws.String(f.eTag),
+		})
+
+	if err != nil {
+		err, ok := err.(awserr.RequestFailure)
```
Comment (jszwec) [inline file.go:139]: please use errors.As instead

File: file.go [hunk 26]
```diff
@@ -1,20 +1,153 @@
...
+		if ok && err.StatusCode() == http.StatusPreconditionFailed {
+			return 0, fs.ErrNotExist
```
Comment (jszwec) [inline file.go:141]: hmm maybe we could wrap this error with some more detail

```go
return 0, fmt.Errorf("s3fs.file.Seek: file has changed while seeking: %w", fs.ErrNotExist)
```

File: file.go (hunk 24 above)
Comment (WilliamFrei) [reply file.go]: Removed, also removed the offset parameter from `getStatFunc`. The function itself still exists but could be inlined as it is only called at a single place now.

File: file.go (hunk 25 above)
Comment (WilliamFrei) [reply file.go]: Changed

File: file.go (hunk 26 above)
Comment (WilliamFrei) [reply file.go]: Changed

File: file.go (hunk 23 above)
Comment (WilliamFrei) [reply file.go:130]: Changed & added the BUG comment.

Was moving the struct below the `New` function intentional?

Comment (jszwec) [reply file.go:130]: > Was moving the struct below the New function intentional?
no, but either works

**Review by jszwec — APPROVED** [review]

**Comment by jszwec** [comment]
LGTM thanks!
